src/shared/sleep-config.c

```diff
@@ -102,6 +102,88 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
         return 0;
 }
 
+int battery_is_low(void) {
+        _cleanup_free_ char *bat_cap_level = NULL;
+        int r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity_level", &bat_cap_level);
+        if (r < 0) {
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity_level: %m");
+        }
+
+        log_debug("Current battery level: %s", bat_cap_level);
+
+        return STR_IN_SET(bat_cap_level, "Critical", "Low");
+}
+
+int read_battery_capacity_percentage(void) {
+        _cleanup_free_ char *bat_cap = NULL;
+        int battery_capacity, r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity", &bat_cap);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity: %m");
+
+        r = safe_atoi(bat_cap, &battery_capacity);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse battery capacity: %m");
+
+        if (battery_capacity < 0 || battery_capacity > 100)
+               return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid battery capacity");
+
+        log_debug("Current battery charge percentage: %d%%", battery_capacity);
+
+        return battery_capacity;
+}
+
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *dirpath = NULL;
+        int stored_discharge_rate, r;
+        char *dis_rate;
+        sd_id128_t machine_id = SD_ID128_NULL;
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get machine-id: %m");
+
+        dirpath = path_join("/var/lib/systemd/sleep", SD_ID128_TO_STRING(machine_id), "battery_discharge_percentage_rate_per_hour");
+        if (!dirpath)
+               return log_oom_debug();
+
+        r = read_one_line_file(dirpath, &dis_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read discharge rate from %s: %m", dirpath);
+
+        r = safe_atoi(dis_rate, &stored_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse discharge rate read from %s location: %m", dirpath);
+
+        if (stored_discharge_rate > 0 && stored_discharge_rate < 200)
+               return stored_discharge_rate;
+        else
```

Comment:
`else` is unnecessary here

Reply:
Should i remove the complete else block?
Not log the error ?

Reply:
No, just the `else` statement is not necessary, as it returns in the `if`, the log is fine